src/backend/optimizer/plan/createplan.c

```diff
@@ -9063,4 +9064,148 @@ push_locus_down_after_elide_motion(Plan* plan)
 				plan = plan->lefttree;
 		}
 	}
-}
\ No newline at end of file
+}
+
+/*
+ * Restore Entry locus to SingleQE in the root slice.
+ * This is simply a reverse of push_locus_down_after_elide_motion.
+ * The difference is that it's NOT used when creating a plan but rather
+ * after a plan gets created, it's used to modify the plan in offload_entry_to_qe.
+ */
+static void
+replace_entry_locus_with_singleqe(Plan *plan)
+{
+	while (plan && (plan->locustype == CdbLocusType_Entry))
+	{
+		plan->locustype = CdbLocusType_SingleQE;
+		switch (nodeTag(plan))
+		{
+			case T_Motion:
+				return;
```

Comment:
since return , but plan->locustype has changed ?

Reply:
this is intended. locus of motion should reflect the receiver side. see https://github.com/cloudberrydb/cloudberrydb/pull/385/files/58a5e51a2a478454b4c9a24b5d1e4d72457ff933#diff-e8ae6e7c9f7b97cc22646b55cd2c66317a4cd34954186bf6d966b80dc085e683R62-R67 as an example.